networks/decoder_net.py

```diff
@@ -0,0 +1,130 @@
+to the input abstract position q. In the paper q is a (16, 4, 4) tensor that can be seen as a 4x4
+image with 16 channels, but here other sizes may be used.
+"""
+
+import torch
+from torch import nn
+
+
+class ResidualBlock(nn.Module):
+    """A residual block that up-samples the input image by a factor of 2.
+    """
+
+    def __init__(self, in_channels, n_filters=64, kernel_size=3, padding='same'):
+        """Instantiate the residual block, composed by a 2x up-sampling and two convolutional
+        layers.
+
+        Args:
+            in_channels (int): Number of input channels.
+            n_filters (int): Number of filters, and thus output channels.
+            kernel_size (int): Size of the convolutional kernels.
+            padding (Union[str, int]): Padding for the convolutional kernels. Can be an int,
+                or 'same' to automatically compute the padding to preserve the image size.
+        """
+        super().__init__()
+        self.channels = in_channels
+        self.n_filters = n_filters
+        if isinstance(padding, int):
+            use_padding = padding
+        elif padding == 'same':
+            use_padding = int(kernel_size / 2)
+        else:
+            raise ValueError('Padding argument not understood. Must be integer or "same".')
+        self.conv1 = nn.Conv2d(
+            in_channels=in_channels,
+            out_channels=n_filters,
+            kernel_size=kernel_size,
+            padding=use_padding,
+            bias=False
+        )
+        self.conv2 = nn.Conv2d(
+            in_channels=n_filters,
+            out_channels=n_filters,
+            kernel_size=kernel_size,
+            padding=use_padding,
+            bias=False
+        )
+        if in_channels != n_filters:
+            self.dim_match_conv = nn.Conv2d(
+                in_channels=in_channels,
+                out_channels=n_filters,
+                kernel_size=1,
+                bias=False,
+                padding=0
+            )
+        self.leaky_relu = nn.LeakyReLU()
+        self.upsample = nn.UpsamplingNearest2d(scale_factor=2)
+        self.sigmoid = nn.Sigmoid()
+
+    def forward(self, x):
+        """Apply 2x up-sampling, followed by two convolutional layers with leaky relu. A sigmoid
+        activation is applied at the end.
+
+        TODO: Should we use batch normalization? It is often common in residual blocks.
+        TODO: Here we apply a convolutional layer to the input up-sampled tensor if its number
+            of channels does not match the convolutional layer channels. Is this the correct way?
+
+        Args:
+            x (torch.Tensor): Input image of shape (N, C, H, W) where N is the batch size and C
+                is the number of in_channels.
+
+        Returns:
+            A torch.Tensor with the upsampled images, of shape (N, n_filters, H, W).
+        """
+        x = self.upsample(x)
+        residual = self.dim_match_conv(x) if self.channels != self.n_filters else x
```

Comment:
Is this a deep copy or a pointer to x?

Reply:
What do you mean? Here I reassign the variable `x` to the upsampled value, and if the number of channels does not match, I store in residual the result of a 1x1 convolution that turns x into a tensor with the correct number of channels.

Reply:
Just wanted to make sure that residual and x where in different things, i guess thats the case sorry